pkg/loadbalancer/flow/flow.go

```diff
@@ -17,50 +17,138 @@ limitations under the License.
 package flow
 
 import (
+	"fmt"
+	"strconv"
+
 	nspAPI "github.com/nordix/meridio/api/nsp/v1"
 	"github.com/nordix/meridio/pkg/loadbalancer/types"
 	"github.com/sirupsen/logrus"
 )
 
+// Flow holds flow data
 type Flow struct {
 	*nspAPI.Flow
 	nfqueueLoadBalancer types.NFQueueLoadBalancer
+	nftHandler types.NftHandler
 }
 
-func New(flow *nspAPI.Flow, options ...Option) (types.Flow, error) {
-	opts := &flowOptions{}
-	for _, opt := range options {
-		opt(opts)
+// New creates a new flow
+func New(flow *nspAPI.Flow, lb types.NFQueueLoadBalancer, nfth types.NftHandler) (types.Flow, error) {
+	if flow == nil {
+		return nil, fmt.Errorf("Flow:New: Create nil Flow")
+	}
+	logrus.Infof("Flow: New %s", flow.Name)
+	if lb == nil {
+		return nil, fmt.Errorf("Flow(%s):New: No NFQueueLoadBalancer", flow.Name)
+	}
+	if nfth == nil {
+		return nil, fmt.Errorf("Flow(%s):New: No NftHandler", flow.Name)
+	}
+
+	if flow.LocalPort != 0 {
+		dport, err := ensureSingleDport(flow)
+		if err != nil {
+			return nil, err
+		}
+		logrus.Infof("Flow:New: Port NAT: %v -> %v", dport, flow.LocalPort)
+
+		err = nfth.PortNATCreateSets(flow)
+		if err != nil {
+			return nil, err
+		}
+		err = nfth.PortNATSetAddresses(flow)
+		if err != nil {
+			nfth.PortNATDeleteSets(flow)
+			return nil, err
+		}
+
+		err = nfth.PortNATSet(
+			flow.Name, flow.Protocols, dport, uint(flow.LocalPort))
+		if err != nil {
+			nfth.PortNATDeleteSets(flow)
+			return nil, err
+		}
 	}
 
 	f := &Flow{
 		Flow:                flow,
-		nfqueueLoadBalancer: opts.nfqueueLoadBalancer,
+		nfqueueLoadBalancer: lb,
+		nftHandler:          nfth,
 	}
-
-	var err error
-	if f.nfqueueLoadBalancer != nil {
-		err = f.nfqueueLoadBalancer.SetFlow(flow)
+	if err := f.nfqueueLoadBalancer.SetFlow(f.Flow); err != nil {
+		if flow.LocalPort != 0 {
+			nfth.PortNATDelete(flow.Name)
+			nfth.PortNATDeleteSets(flow)
+		}
+		return nil, err
 	}
-	logrus.Infof("New flow: %v - err: %v", flow, err)
 	return f, nil
 }
 
+// Update updates a flow
 func (f *Flow) Update(flow *nspAPI.Flow) error {
-	if f.Flow == nil || !f.Flow.DeepEquals(flow) {
-		f.Flow = flow
-		logrus.Infof("Update flow: %v", f.Flow)
-		if f.nfqueueLoadBalancer != nil {
-			return f.nfqueueLoadBalancer.SetFlow(f.Flow)
+	logrus.Infof("Flow:Update: %v -> %v", f.Flow, flow)
```

Comment:
This will flood the logs. Any change leads to pushing all the flows to the lb. (Including NSP failover.)

Reply:
Changed to trace level